### PR TITLE
Adv/file formatting

### PR DESCRIPTION
**Pull request by antonydellavecchia**
Reformatting polynomial serialization for better efficiency
 the terms of a polynomial are now stored as
   terms = [(exponent, coeff), ..., (exponent, coeff)]
 where exponent is either an int or a vector of ints, and coeff is either a string
 or is given as a vector of terms. This results in a nested structure, the parents
 for the coefficients are found in the parents vector. The parent of the
coefficient can be determined based on the depth level of the given coefficient
 This new format also allows us to improve serialization of vectors of ring
elements, since the parents vector can be brought to the root, and we use
 the vector entries to store the terms

### CODE REVIEW

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl]: Instead of forcing calls that pass a value for `str` to also pass one for `uses_id`, even if it is false, I think you can use "macro overloading: I.e. have two macros with the same name but different argument lists:
```julia-repl
julia> macro foo1(ex::Any, uses_id::Bool = false, str::Union{String,Nothing} = nothing)
         return :(:bar, $ex, $uses_id, $str)
       end
@foo1 (macro with 3 methods)

julia> @foo1 1+2
(:bar, 3, false, nothing)

julia> macro foo1(ex::Any, str::Union{String,Nothing})
         return :(:quux,$ex,false,$str)
       end
@foo1 (macro with 4 methods)

julia> @foo1 1+2
(:bar, 3, false, nothing)

julia> @foo1 1+2 false
(:bar, 3, false, nothing)

julia> @foo1 1+2 "house"
(:quux, 3, false, "house")
```

Comment (fingolfin) [reply src/Serialization/main.jl]: Oh, and those two macro implementations can share code: just move the bulk of the macro implementation into a regular function, then invoke it from both macro variants

**Comment by antonydellavecchia** [comment]
closes https://github.com/oscar-system/Oscar.jl/issues/2190

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl]:
```suggestion
            serialize_with_id(obj::T) where T <: $ex = $uses_id && !has_elem_basic_encoding(obj)
```

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl]: Maybe add a comment here. 
```suggestion
    # For rings whose elements have basic encoding (e.g. Z/7Z), we don't bother
    # storing backrefs to the ring, we just create a new instance of the ring.
    if has_elem_basic_encoding(obj) && s.depth != 0
```

Also: when can it happens that `s.depth == 0`? That would mean we invoke `save_as_ref` as the "top level", no? And anyway, how can we then store a backref if we created all those rings without an id ?!?!?

Maybe we can either give an example (in a comment, and/or a test); or else argue that it cannot happen (and then we just put in an `@assert s.depth != 0` or something like this)

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl:218]: So depth 0 is somewhat special. Maybe we need some test cases just for it, to make sure things work right?

File: src/Serialization/Fields.jl
Comment (fingolfin) [inline src/Serialization/Fields.jl]: Strictly speaking instead of `length(gens(x))` one should do `ngens(x)`, *but* (a) I have not verified it exists in those case (I mean: it *should* but it might be one of the many places that we need to fix...); and (b) of course then the local variable shouldn't be called `ngens` anymore but e.g. `n`

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl:65]: Actually it is more complex now, because also `has_elem_basic_encoding` plays a role... 

I am wondering now if we should rethink the relation between `has_elem_basic_encoding` and `serialize_with_id` and this macro... Like, when I specify that "elements have basic encoding", then we automatically don't store the id. Only if element do *not* have basic encoding is it even an option to store the id. Perhaps we can come up with a slightly less complex way to specify all of this?

File: src/Serialization/main.jl
Comment (fingolfin) [reply src/Serialization/main.jl]: Actually, now I wonder: shouldn't this use `serialize_with_id` here? I am confused, sorry :-)

File: src/Serialization/Fields.jl
Comment (fingolfin) [inline src/Serialization/Fields.jl]: I am slighty concerned that we kind of hard-code here the requirement of first producing nested dictionaries, and then serializing them at the end. I still hope that we can arrange the code in a way that allows for other serialization formats which are e.g. more suitable for inter process communication (see issue #1507).

Of course we are already constructing a `Dict` in this function! But this wouldn't be too hard to change (see sketch in #1507). But if we now rely on the return value to be dict...

So maybe we can extract part of function which serializes the `Vector` into a new helper which is called from here (say `save_vector` or so...) and then just also call this from `save_internal(s, ::Vector,...)`

File: src/Serialization/containers.jl
Comment (fingolfin) [inline src/Serialization/containers.jl]: Is this implicitly relying on the fact that `serialize_with_id` will return `false` for this parent? Or what's going on here?

It seems that `has_elem_basic_encoding(X)` has two meanings at once:
1. the element of `X` have a "basic encoding"
2. `X` itself can be encoded in a "basic" fashion, namely we don't need to encode any of its baserings etc. ?

I think it might help to separate these two concepts into two traits (even if in the end they will always have equal value)

File: src/Serialization/Rings.jl
Comment (fingolfin) [inline src/Serialization/Rings.jl]: I don't quite understand what's going on here. Why isn't it
```suggestion
    if has_elem_basic_encoding(parent_ring)
        return Any[]
    end
    base = base_ring(parent_ring)
```

Comment (fingolfin) [reply src/Serialization/Rings.jl]: ... and then it would be basically identical to this one from above (so that method could be removed)
```julia

 function get_parents(parent_ring::T) where T <: Union{FracField,
                                                       AbstractAlgebra.Generic.RationalFunctionField,
                                                       AbstractAlgebra.Generic.LaurentSeriesField}
     if has_elem_basic_encoding(parent_ring)
         return Any[]
     end

     base = base_ring(parent_ring)
     parents = get_parents(base)
     push!(parents, parent_ring)
     return parents
 end
```

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl]: I think before we merge this PR, these definitions should be merged next to the corresponding `@registerSerializationType` calls.

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl:168]: We need some comments that explain the (intended) semantics for `has_elem_basic_encoding` and `has_basic_encoding` and in particular clarify how they differ from `is_basic_serialization_type` ?

File: src/Serialization/Rings.jl
Comment (fingolfin) [inline src/Serialization/Rings.jl]:
```suggestion
        if iszero(coeff(p, i))
```
Does this ever happen?

File: src/Serialization/Rings.jl
Comment (fingolfin) [inline src/Serialization/Rings.jl]: Before the loop:
```julia
    parents = save_parents_internal(s, SOMETHING)
```
then here
```suggestion
        # no need to serialize the parents here, we already did;
        # in this case, the serialization of the coefficient does not
        # need to store the parents; and also it can directly store (return) the terms, instead of putting them into a sublist
        encoded_coeff = save_internal(s, coeff(p, i); omit_parents=true)
        push!(encoded_terms, (exponent_vector(p, i), encoded_coeff))
```

File: src/Serialization/Rings.jl
Comment (fingolfin) [inline src/Serialization/Rings.jl]: Maybe this?
```suggestion
                c = base(coeff)
```

Comment (fingolfin) [reply src/Serialization/Rings.jl]: Or if you can't get here, maybe add an `@assert` or so?

File: src/Serialization/Rings.jl
Comment (fingolfin) [inline src/Serialization/Rings.jl]:
```suggestion
    for e, coeff in terms
```

File: src/Serialization/main.jl
Comment (antonydellavecchia) [reply src/Serialization/main.jl]: this now checks if serialize with id is false and returns the encoded obj, storing of any  parent types now go through a save_ref, maybe this means the function deserves a name change to save_parent

File: src/Serialization/Fields.jl
Comment (antonydellavecchia) [reply src/Serialization/Fields.jl]: save_internal for vector now has an include_parents flag

File: src/Serialization/Rings.jl
Comment (antonydellavecchia) [reply src/Serialization/Rings.jl]: this never happens and was removed

File: src/Serialization/Rings.jl
Comment (antonydellavecchia) [reply src/Serialization/Rings.jl]: resolved by adding an include_parents flag for save_internal of any ring elements, we use the get_parent_refs function now if the flag is true and returns a dict with a parents key, otherwise just returns the terms as a vector

File: src/Serialization/Rings.jl
Comment (antonydellavecchia) [reply src/Serialization/Rings.jl]: line 208 could never be reached

File: src/Serialization/main.jl
Comment (antonydellavecchia) [reply src/Serialization/main.jl:65]: I've opted to keep these two ideas seperate to avoid confusion or possible inconsistences in the future. i.e. Objects whose elements have basic encodings may also be stored with an id. Having elements with basic encoding just means that the object has no base_ring, and represents a sort of atomic fundamental type at least within oscar.

Comment (antonydellavecchia) [reply src/Serialization/main.jl:65]: I've also removed the dependancy to the has_elem_basic_encoding

File: src/Serialization/containers.jl
Comment (antonydellavecchia) [reply src/Serialization/containers.jl]: I've added a note above has_elem_basic_encoding definition, to clarify what is meant and how it is used

File: src/Serialization/main.jl
Comment (antonydellavecchia) [reply src/Serialization/main.jl:218]: tests have been added

**Comment by antonydellavecchia** [comment]
closes https://github.com/oscar-system/Oscar.jl/issues/2191

**Comment by antonydellavecchia** [comment]
closes https://github.com/oscar-system/Oscar.jl/issues/2193

**Comment by antonydellavecchia** [comment]
closes https://github.com/oscar-system/Oscar.jl/issues/2513
closes https://github.com/oscar-system/Oscar.jl/issues/2514

File: src/Serialization/Fields.jl
Comment (fingolfin) [inline src/Serialization/Fields.jl]:
```suggestion
                       ::Type{<:FracElem},
```

File: src/Serialization/Fields.jl
Comment (fingolfin) [inline src/Serialization/Fields.jl:66]: This pattern repeats over and over again. I think there is an opportunity to factor out this duplication (possibly in a future PR)...

Is ther already a trait function `is_FOO` which returns true for all types that honor `include_parents` (so like `fpFieldElem` here)? If not, we could add one, that would help with this refactoring. 

And then I think we could handle it uniformly at the top level (just like you already handle it uniformly when loading, I think), by editing `save_type_dispatch`: just replace  `foo = save_internal(s, obj)` by something like
```julia
   if is_type_storing_parents(typeof(obj))
      p = save_as_ref(s, parent(obj))
      foo = Dict(:parent => p, :data => save_internal(s, obj))
   else
      foo = save_internal(s, obj)
   end
```
Of course if instead of the generic `:data` we want something more bespoke like `:class_rep`, that's also possible (e.g. add a function ```
tag_for_serialization_data(::Any) = :data
tag_for_serialization_data(::fpFieldElem) = :class_rep
```
and use that appropriately.

**Review by benlorenz — APPROVED** [review]
This is already a great improvement and implements many of the requested features. I did not see any major issues in the code and I think to move forward we should merge this as soon as possible and defer more changes to future PRs.

PS: nightly failures are unrelated.

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl:165]: The relation between `is_basic_serialization_type`, `has_elem_basic_encoding` and `has_basic_encoding` still is confusing to me; and the latter two names are so similar that I keep misreading one for the other. But I don't want to hold up this PR for that, I'll try and see if I can come up with better alternatives, this can still be changed in the future.

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl:135]: I am still not happy with `is_basic_serialization_type` as a name but I don't want to hold up this PR further, so I'll think about a better name later.

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl:66]:
```suggestion
# whether or not the type can be back referenced.
# If omitted, the default is that no back references are allowed.
```

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl]:
```suggestion
            # There exist types where equality cannot be discerned from the serialization
```

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl:79]: I don't understand the preceding two lines. Perhaps an example would be illustrative.. or perhaps two? one for when this is needed and one when not?

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl:91]: So is the "id" the same as the "uuid"?

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl]: "Objects have a elements" ??

Is this only talking about "things that have elements" (like rings, groups, ...) or about anything that has basic serialization?

File: src/Serialization/main.jl
Comment (fingolfin) [inline src/Serialization/main.jl:151]: So... "basic encoding" = "basic serialization"? If so, perhaps we should align the names? If not, i.e. there is a difference here: what is it?

**Review by fingolfin — COMMENTED** [review]
I think there are great improvements here, but there is still quite a bit to be done. Unfortunately I don't have time right now to give this the full review it deserves.

Since @micjoswig and @antonydellavecchia want to work on this, feel free to merge it.

The one drawback is that then it gets harder to review these changes  (unless one goes back to this PR), but so be it.

File: src/Serialization/main.jl
Comment (antonydellavecchia) [reply src/Serialization/main.jl:91]: yes

File: src/Serialization/main.jl
Comment (antonydellavecchia) [reply src/Serialization/main.jl:151]: i've added some more details describing this to the code. Basically the difference is that encodings rely on a parent.
so some basic encodings use basic serialization, elements of ZZ for example, however, numbers of type Int64 dont really have an encoding since they don't really have a parent, same for the type Symbol, Bool. We also have ZZ/nZZ whose elements have a basic encoding but not a basic serialization, we cannot just store "6" and know it represents "6" in ZZ/7ZZ, we need to know the parent

**Pull request merged**
